org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/IndexManager.java

```diff
@@ -251,10 +266,10 @@ public void addSource(IFile resource, IPath containerPath, SourceElementParser p
 	IndexLocation indexLocation = computeIndexLocation(containerPath);
 	scheduleDocumentIndexing(document, containerPath, indexLocation, participant);
 }
-/*
+/**
  * Removes unused indexes from disk.
  */
-public void cleanUpIndexes() {
+public synchronized void cleanUpIndexes() {
```

Comment:
Note: this is called from `org.eclipse.core.internal.resources.SaveManager.save()` that will lock the workspace before. Means, we might block workspace save and in worst case we have other code that locks the index manager before and needs workspace lock to do something, just like the code few lines below calling to `BasicSearchEngine.createWorkspaceScope()`.

Reply:
OK, I've checked now the callees of the code inside, the change should be safe.
```
cleanUpIndexes() ->  synchronized
	conditionally called removeIndexesState() (synchronized)
	called getIndexes(IndexLocation[], IProgressMonitor) (non synchronized)
		called getIndex(IndexLocation) (synchronized)
		called conditionally getIndex(IPath, IndexLocation, boolean, boolean) (synchronized)
		called conditionally updateIndexState(IndexLocation, Integer) (synchronized)
```